Current errors

```
❯ make
clang++ -g -Wall -std=c++11 -DDEBUG -Isrc -Itest/lib -o run-tests test/main.cpp
In file included from test/main.cpp:1:
In file included from src/roc_std.h:8:
src/result.h:30:13: error: multiple overloads of 'ResultPayload' instantiate to the same signature 'void (const Roc::Str &)'
            ResultPayload(const E &other)
            ^
src/result.h:38:11: note: in instantiation of member class 'Roc::Result<Roc::Str, Roc::Str>::ResultPayload' requested here
        } m_payload;
          ^
test/result.cpp:9:33: note: in instantiation of template class 'Roc::Result<Roc::Str, Roc::Str>' requested here
    const Result<Str, Str> ok = Result<Str, Str>::make_ok(str);
                                ^
src/result.h:24:13: note: previous declaration is here
            ResultPayload(const O &other)
            ^
1 error generated.
make: *** [run-tests] Error 1
```

--- a/README.md
+++ b/README.md
@@ -20,4 +20,4 @@ Current progress
 - [ ] Dict
 - [ ] Set
 - [x] Box
-- [ ] Result
+- [x] Result

--- a/src/num.h
+++ b/src/num.h
@@ -5,10 +5,11 @@
 namespace Roc
 {
     template <typename T>
-    class Num : Value
+    class Num : public Value
     {
     public:
         T value;
+        Num() : value(0) {}
         Num(T value) : value(value) {}
 
         bool operator== (const Num<T> &other) const

--- a/src/result.h
+++ b/src/result.h
@@ -1,0 +1,75 @@
+#pragma once
+
+#include "internal.h"
+
+namespace Roc
+{
+    template <typename O, typename E>
+    class Result : Value
+    {
+        union Payload
+        {
+            O ok;
+            E err;
+        } m_payload;
+        bool m_is_ok;
+
+    public:
+        static Result make_ok(O val)
+        {
+            Result r;
+            r.m_payload.ok = val;
+            r.m_is_ok = true;
+            return r;
+        }
+
+        static Result make_err(E val)
+        {
+            Result r;
+            r.m_payload.err = val;
+            r.m_is_ok = false;
+            return r;
+        }
+
+        bool is_ok() const
+        {
+            return m_is_ok;
+        }
+
+        O* ok() const
+        {
+            return m_is_ok ? &m_payload->ok : NULL;
+        }
+
+        E* err() const
+        {
+            return m_is_ok ? NULL : &m_payload->err;
+        }
+
+        ~Result()
+        {
+            rc_decrement();
+        }
+
+        bool rc_unique() const
+        {
+            return true;
+        };
+
+        void rc_increment()
+        {
+            if (m_is_ok)
+                m_payload.ok.rc_increment();
+            else
+                m_payload.err.rc_increment();
+        };
+
+        void rc_decrement()
+        {
+            if (m_is_ok)
+                m_payload.ok.rc_decrement();
+            else
+                m_payload.err.rc_decrement();
+        };
+    };
+};

--- a/src/result.h
+++ b/src/result.h
@@ -113,4 +113,72 @@ namespace Roc
                 m_payload.err.rc_decrement();
         };
     };
+
+    // Specialization for when Ok and Error are the same type
+    template <typename T>
+    class Result<T, T> : Value
+    {
+        T m_payload;
+        bool m_is_ok;
+
+    public:
+        Result() {}
+
+        Result(const Result &other)
+            : m_payload(other.m_payload), m_is_ok(other.m_is_ok)
+        {
+            rc_increment();
+        }
+
+        static Result make_ok(T val)
+        {
+            Result r;
+            r.m_payload = val;
+            r.m_is_ok = true;
+            return r;
+        }
+
+        static Result make_err(T val)
+        {
+            Result r;
+            r.m_payload = val;
+            r.m_is_ok = false;
+            return r;
+        }
+
+        bool is_ok() const
+        {
+            return m_is_ok;
+        }
+
+        T *ok() const
+        {
+            return m_is_ok ? &m_payload : NULL;
+        }
+
+        T *err() const
+        {
+            return m_is_ok ? NULL : &m_payload;
+        }
+
+        ~Result()
+        {
+            m_payload.~T();
+        }
+
+        bool rc_unique() const
+        {
+            return true;
+        };
+
+        void rc_increment()
+        {
+            m_payload.rc_increment();
+        };
+
+        void rc_decrement()
+        {
+            m_payload.rc_decrement();
+        };
+    };
 };

--- a/src/result.h
+++ b/src/result.h
@@ -26,7 +26,7 @@ namespace Roc
                 ok = other;
             }
 
-            // This is a problem if O and E are the same type            
+            // This is a problem if O and E are the same type
             ResultPayload(const E &other)
             {
                 err = other;
@@ -45,9 +45,8 @@ namespace Roc
         }
 
         Result(const Result &other)
+            : m_payload(other.m_payload), m_is_ok(other.m_is_ok)
         {
-            m_is_ok = other.m_is_ok;
-            m_payload = other.m_payload;
             rc_increment();
         }
 

--- a/src/roc_std.h
+++ b/src/roc_std.h
@@ -6,3 +6,4 @@
 #include "box.h"
 #include "num.h"
 #include "dict.h"
+#include "result.h"

--- a/test/main.cpp
+++ b/test/main.cpp
@@ -7,6 +7,7 @@
 #include "list.cpp"
 #include "box.cpp"
 #include "dict.cpp"
+#include "result.cpp"
 
 int main(int argc, char **argv)
 {

--- a/test/result.cpp
+++ b/test/result.cpp
@@ -16,9 +16,6 @@ namespace Roc
         const Result<I64, Str> err = Result<I64, Str>::make_err(str);
         TEST(!err.is_ok());
     }
-    /*
-
-    // Won't compile if O and E are the same type
 
     DEFINE_TEST_G(OkSameType, Result)
     {
@@ -33,6 +30,4 @@ namespace Roc
         const Result<Str, Str> err = Result<Str, Str>::make_err(str);
         TEST(!err.is_ok());
     }
-
-    */
 };

--- a/test/result.cpp
+++ b/test/result.cpp
@@ -1,0 +1,20 @@
+#include "roc_std.h"
+#include "lib/simpletest.h"
+
+namespace Roc {
+
+DEFINE_TEST_G(Ok, Result)
+{
+    const Str str("It's all good, no errors in this long string!");
+    const Result<Str, Str> ok = Result<Str, Str>::make_ok(str);
+    TEST(ok.is_ok());
+}
+
+DEFINE_TEST_G(Err, Result)
+{
+    const Str str("Oops, this long string is about a nasty error!");
+    const Result<Str, Str> err = Result<Str, Str>::make_err(str);
+    TEST(!err.is_ok());
+}
+
+};

--- a/test/result.cpp
+++ b/test/result.cpp
@@ -1,20 +1,38 @@
 #include "roc_std.h"
 #include "lib/simpletest.h"
 
-namespace Roc {
-
-DEFINE_TEST_G(Ok, Result)
+namespace Roc
 {
-    const Str str("It's all good, no errors in this long string!");
-    const Result<Str, Str> ok = Result<Str, Str>::make_ok(str);
-    TEST(ok.is_ok());
-}
 
-DEFINE_TEST_G(Err, Result)
-{
-    const Str str("Oops, this long string is about a nasty error!");
-    const Result<Str, Str> err = Result<Str, Str>::make_err(str);
-    TEST(!err.is_ok());
-}
+    DEFINE_TEST_G(Ok, Result)
+    {
+        const Str str("It's all good, no errors in this long string!");
+        const Result<Str, I64> ok = Result<Str, I64>::make_ok(str);
+        TEST(ok.is_ok());
+    }
+    DEFINE_TEST_G(Err, Result)
+    {
+        const Str str("Oops, this long string is about a nasty error!");
+        const Result<I64, Str> err = Result<I64, Str>::make_err(str);
+        TEST(!err.is_ok());
+    }
+    /*
 
+    // Won't compile if O and E are the same type
+
+    DEFINE_TEST_G(OkSameType, Result)
+    {
+        const Str str("It's all good, no errors in this long string!");
+        const Result<Str, Str> ok = Result<Str, Str>::make_ok(str);
+        TEST(ok.is_ok());
+    }
+
+    DEFINE_TEST_G(ErrSameType, Result)
+    {
+        const Str str("Oops, this long string is about a nasty error!");
+        const Result<Str, Str> err = Result<Str, Str>::make_err(str);
+        TEST(!err.is_ok());
+    }
+
+    */
 };

